pkg/controller/csr/certificatesigningrequest.go

```diff
@@ -0,0 +1,225 @@
+			Debugf("CSR already approved/denied")
+		return ret, nil
+	}
+
+	if cnt := len(csr.Status.Certificate); cnt > 0 {
+		log.Debugf("CSR Certificate already signed  (%d certs attached)", cnt)
+		return ret, nil
+	}
+
+	// ignore no labels, or missing label name=portworx, or missing label cid
+	var clusterID string
+	if labs := csr.GetLabels(); labs == nil {
+		log.Debugf("Ignoring CSR  (no labels)")
+		return ret, nil
+	} else if labs[labelNameKey] != labelNameVal {
+		log.Debugf("Ignoring CSR  (no label %s=%s)", labelNameKey, labelNameVal)
+		return ret, nil
+	} else if clusterID = labs[labelClusterIDKey]; len(clusterID) <= 0 {
+		log.Debugf("Ignoring CSR  (no label %s)", labelClusterIDKey)
+		return ret, nil
+	}
+
+	// check registration, requeue if not (yet) registered, otherwise approve or ignore
+	isApprove, has := registration.Load(clusterID)
+	if !has {
+		log.Debugf("Postponing CSR eval  (clusterID %s not registered)", clusterID)
+		return ctrl.Result{Requeue: true}, nil
+	} else if !isApprove.(bool) {
+		log.WithField("clusterID", clusterID).
+			Warnf("NOT auto-approving CSR -- approve manually via `kubectl certificate approve %s`)", req.Name)
+		return ret, nil
+	}
+
+	// CHECKME: should we inspect CSR (e.g. check DNS names + IPs, expiration)
+
+	log.WithField("clusterID", clusterID).Info("Approving CSR")
+	c.appendCondition(&csr, true, "")
+
+	_, err := coreops.Instance().CertificateSigningRequestsUpdateApproval(req.Name, &csr)
+	if apierrors.IsConflict(err) || apierrors.IsNotFound(err) {
+		// The CSR has been updated or deleted since we read it.
+		// Requeue the CSR to try to reconcile again.
+		return ctrl.Result{Requeue: true}, nil
+	} else if err != nil {
+		log.WithError(err).Errorf("Error approving CSR")
+		return ret, err
+	}
+
+	return ret, nil
+}
+
+// getCertApprovalCondition scans CSR status, returns <approved> and <denied>
+func (c *Controller) getCertApprovalCondition(st *certv1.CertificateSigningRequestStatus) (approved, denied bool) {
+	for _, cond := range st.Conditions {
+		switch cond.Type {
+		case certv1.CertificateApproved:
+			approved = true
+		case certv1.CertificateDenied:
+			denied = true
+		}
+	}
+	return
+}
+
+// appendCondition updates CSR status as approved or denied
+func (c *Controller) appendCondition(csr *certv1.CertificateSigningRequest, approved bool, reason string) {
+	cnd := certv1.CertificateSigningRequestCondition{
+		Type:               certv1.CertificateDenied,
+		Status:             corev1.ConditionTrue,
+		Reason:             "cert denied by px-operator",
+		Message:            "CSR failed px-operator validation process. Reason: " + reason,
+		LastUpdateTime:     metav1.Now(),
+		LastTransitionTime: metav1.Time{},
+	}
+	if approved {
+		cnd.Type = certv1.CertificateApproved
+		cnd.Reason = "cert validated by px-operator"
+		cnd.Message = "CSR approved by px-operator"
+	}
+	csr.Status.Conditions = append(csr.Status.Conditions, cnd)
```

Comment:
nit: Should we check if the condition of that type already exists and edit that instead of directly appending it?

Reply:
Note -- we don't actually call the `appendCondition()` "unconditionally" --
* if CSR has been approved/denied, we'll exit `Reconcile()`

ref from `Reconcile():137`
```go
	if approved, denied := c.getCertApprovalCondition(&csr.Status); approved || denied {
		log.WithField("approved", approved).WithField("denied", denied).
			Debugf("CSR already approved/denied")
		return ret, nil
	}
```